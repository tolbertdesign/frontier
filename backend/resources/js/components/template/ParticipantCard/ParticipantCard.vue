<template>
  <div class="participant-card max-w-xl m-4">
    <div>
      <div class="media mb-2 lg:px-4 relative">
        <div class="media-left">
          <a @click="editParticipant(participant)">
            <AvatarImage
              :alt="participant.first_name"
              :src="avatarSrc"
            />
          </a>
        </div>
        <div class="media-content self-center">
          <h2 class="text-xl">
            {{ participant.first_name }}
          </h2>
        </div>
        <div class="media-right self-start has-text-right">
          <a
            class="text-sm has-text-grey"
            @click="editParticipant(participant)"
          >
            {{ lang.participant_card.edit_profile_link }}
          </a>
        </div>
      </div>
    </div>
    <section class="lg:px-4">
      <div
        class="mb-4"
        style="display: block"
      >
        <ProgressMeter
          :first-meter="totalPledged"
          :second-meter="totalPending"
          :goal="participant.profile.pledge_goal"
        />

        <div class="mt-2">
          <h3 class="font-bold flex items-center">
            <svg
              class="fill-current text-primary h-4 w-4 mr-1"
              viewBox="0 0 16 16"
            >
              <circle
                cx="8"
                cy="8"
                r="8"
              />
            </svg>
            <span>
              {{ lang.participant_card.total_pledged }} ${{ totalPledged | currency }} {{ lang.participant_card.of }} ${{ participant.profile.pledge_goal | currency }}
            </span>
          </h3>
          <h4
            v-if="!isFlat"
            class="text-sm capitalize"
          >
            {{ program.unit.modifier }} {{ program.unit.name }}
          </h4>
        </div>
        <div
          v-if="totalPending > 0"
          class="mt-2"
        >
          <h3 class="font-bold flex items-center">
            <svg
              class="fill-current text-tertiary h-4 w-4 mr-1"
              viewBox="0 0 16 16"
            >
              <circle
                cx="8"
                cy="8"
                r="8"
              />
            </svg>
            <span>
              {{ lang.participant_card.awaiting_confirmation }} ${{ totalPending | currency }}
            </span>
          </h3>
          <h4
            v-if="!isFlat"
            class="text-sm capitalize"
          >
            {{ program.unit.modifier }} {{ program.unit.name }}
          </h4>
        </div>
      </div>
      <div
        :class="[ selected == participant.id ? '' : 'border-b lg:border-b-0' ]"
        class="button-container text-center pb-4"
      >
        <button
          class="py-2 px-8 rounded-full font-semibold text-sm text-white bg-secondary shadow focus:outline-none focus:shadow-outline"
          @click="$emit('toggle', participant.id)"
        >
          <span>{{ selected == participant.id ? lang.close : lang.view }} {{ lang.rewards }}</span>
        </button>
      </div>
    </section>
  </div>
</template>

<script>
import AvatarImage from '@/components/element/AvatarImage'
import ProgressMeter from '@/components/element/ProgressMeter'
import PledgeTotalCalculator from '@/utilities/PledgeTotalCalculator.js'

export default {
  name: 'ParticipantCard',
  status: 'prototype',
  release: '1.0.0',
  components: {
    AvatarImage,
    ProgressMeter,
  },
  props: {
    program: {
      type: Object,
      default: () => ({
        program_pledge_setting: {
          flat_donate_only: 0,
        },
        unit: {
          per: null,
          modifier: null,
        },
        unit_flat_conversion: null,
      }),
    },
    participant: {
      type: Object,
      default: null,
    },
    selected: {
      type: Number,
      default: 0,
    },
    group: {
      type: Number,
      default: 0,
    },
  },

  data () {
    return {
      pledgeStatuses: [
        {
          name: 'confirmed',
          id: 2,
        },
        {
          name: 'paid',
          id: 3,
        },
        {
          name: 'pending',
          id: 4,
        },
        {
          name: 'paymentScheduled',
          id: 8,
        },
      ],
    }
  },

  computed: {
    lang () {
      return this.$store.state.lang
    },
    avatarSrc () {
      return this.participant.profile.image_name == null ? '/v3-assets/dashboard/images/dashboard-avatar.svg' : this.avatarImgSrc
    },
    avatarImgSrc () {
      return this.$store.getters.avatarPath + this.participant.profile.image_name
    },
    editProfileLink () {
      return '/v3/tkdashboard/?redirect=' + encodeURI('auth/login/' + this.participant.fr_code + '/edit-personalize')
    },
    isFlat () {
      return this.program.program_pledge_setting.flat_donate_only === 1
    },
    totalPending () {
      const statusNames = ['pending']
      const statusIds = this.pledgeStatuses
        .filter(status => statusNames.includes(status.name))
        .map(status => status.id)
      return this.getPledgeTotal(statusIds)
    },
    totalPledged () {
      const statusNames = ['confirmed', 'paid', 'paymentScheduled']
      const statusIds = this.pledgeStatuses
        .filter(status => statusNames.includes(status.name))
        .map(status => status.id)
      return parseFloat(this.getPledgeTotal(statusIds))
    },
  },

  methods: {
    getPledgeTotal (statusIds) {
      if (this.program.unit_flat_conversion <= 0) {
        return 0
      }
      if (this.isFlat) {
        return this.participant.participant_info.pledges
          .filter(pledge => statusIds.includes(pledge.pledge_status_id))
          .reduce((total, pledge) => total + pledge.amount, 0)
      }
      const calculator = new PledgeTotalCalculator(this.program, this.participant.participant_info.pledges)
      return calculator.getTotalAsPPL()
    },
    editParticipant (participant) {
      this.$router.push({
        name: 'edit-participant',
        params: {
          id: participant.id,
        },
      })
    },
  },
}
</script>
