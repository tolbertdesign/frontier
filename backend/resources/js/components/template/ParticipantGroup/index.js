import ParticipantGroup from './ParticipantGroup.vue'

export default ParticipantGroup
