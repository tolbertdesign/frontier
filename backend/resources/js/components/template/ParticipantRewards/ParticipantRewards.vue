<template>
  <section class="participant-rewards">
    <Transition name="fade">
      <article
        v-if="show"
        class="py-8"
      >
        <a
          class="text-2xl text-black float-right"
          @click="toggleRewards(participant.id)"
        >
          <i class="fal fa-times" />
        </a>
        <h1 class="text-2xl font-bold">
          {{ lang.rewards }}
        </h1>
        <h2
          v-if="program.program_pledge_setting.flat_donate_only === '0'"
          class="mb-4"
        >
          ${{ program.unit_flat_conversion }} {{ lang.flat_donation }} = $1 {{ program.unit.modifier }} {{ program.unit.name }}
        </h2>

        <TeacherIncentives v-if="shouldDisplayIncentives" />

        <!-- Shown when greater than small, hidden when small -->
        <div class="hidden sm:block mb-4">
          <ParticipantRewardsCard
            v-for="reward in rewards"
            :key="reward.id"
            :reward="reward"
            :program="program"
            :sum-pledges-flat="sumPledgesFlat"
          />
        </div>

        <!-- Shown when small, hidden when greater than small -->
        <div class="sm:hidden mb-4">
          <ParticipantRewardsCard
            v-for="reward in topRewards"
            :key="reward.id"
            :reward="reward"
            :program="program"
            :sum-pledges-flat="sumPledgesFlat"
          />

          <ParticipantRewardsCard
            v-for="reward in remainingRewards"
            :key="reward.id"
            :reward="reward"
            :program="program"
            :sum-pledges-flat="sumPledgesFlat"
            :class="[viewMore ? '' : 'hidden']"
          />
        </div>

        <div class="sm:hidden mb-6 pb-4 text-center border-grey">
          <a
            class="text-blue hover:text-blue-dark"
            @click="toggleViewMore"
          >
            {{ lang.view }} {{ viewMore ? lang.less : lang.more }}
          </a>
        </div>

        <div class="flex items-center justify-center border-b lg:border-none pb-6">
          <button
            class="py-2 px-8 rounded-full font-semibold text-sm text-white bg-secondary shadow focus:outline-none focus:shadow-outline"
            @click="toggleRewards(participant.id)"
          >
            <span>{{ lang.close }} {{ lang.rewards }}</span>
          </button>
        </div>
      </article>
    </Transition>
  </section>
</template>

<script>
import ParticipantRewardsCard from '@/components/template/ParticipantRewardsCard'
import TeacherIncentives from '@/components/template/TeacherIncentives'
import PledgeTotalCalculator from '@/utilities/PledgeTotalCalculator.js'
import { STATUS_CONFIRMED, STATUS_PAID, STATUS_PAYMENT_SCHEDULED, STATUS_PENDING } from '@/store/modules/pledge.js'

export default {
  name: 'ParticipantRewards',
  status: 'prototype',
  release: '1.0.0',
  components: {
    ParticipantRewardsCard,
    TeacherIncentives,
  },
  props: {
    program: {
      type: Object,
      default: null,
    },
    participant: {
      type: Object,
      default: null,
    },
    show: {
      type: Boolean,
      default: false,
    },
  },
  data () {
    return {
      viewMore: false,
    }
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
    shouldDisplayIncentives () {
      if (this.$store.state.User.is_teacher_user && !this.program.hide_teacher_incentives) {
        return this.isTeacherInProgram
      }
      return false
    },
    isTeacherInProgram () {
      if (this.program.participants.find(participant => participant.id === this.$store.state.User.teacher_participant_id)) {
        return true
      }

      return false
    },
    rewards () {
      let combinedPrizeObjects = this.getCombinedPrizeObjects()
      combinedPrizeObjects = this.filterPrizesForDisplay(combinedPrizeObjects)
      combinedPrizeObjects = this.sortPrizesForDisplay(combinedPrizeObjects)
      return combinedPrizeObjects
    },
    topRewards () {
      return this.rewards.slice(0, 4)
    },
    remainingRewards () {
      return this.rewards.slice(4)
    },
    sumPledgesFlat () {
      const calculator = new PledgeTotalCalculator(this.program, this.participant.participant_info.pledges, [
        STATUS_CONFIRMED,
        STATUS_PAID,
        STATUS_PAYMENT_SCHEDULED,
        STATUS_PENDING,
      ])
      return calculator.getTotalAsFlat()
    },
  },
  methods: {
    sortPrizesForDisplay (combinedPrizeObjects) {
      return combinedPrizeObjects.sort((a, b) => {
        return a.prizeBound.display_amount - b.prizeBound.display_amount
      })
    },
    filterPrizesForDisplay (combinedPrizeObjects) {
      return combinedPrizeObjects.filter(prizeBoundStudent => {
        const isNotActivityReward = prizeBoundStudent.prizeBound.activity_reward === null
        const hasNoPeriod = prizeBoundStudent.prizeBound.starts_at === null &&
                    prizeBoundStudent.prizeBound.ends_at === null
        const isNotQuantity = prizeBoundStudent.prizeBound.quantity === null
        let isInPeriod = false
        let isDisplayableStatus = true
        const displayableStatuses = [
          'giveaway',
          'delivered',
          'pending',
        ]
        if (!hasNoPeriod) {
          isInPeriod = new Date(prizeBoundStudent.prizeBound.starts_at) < new Date() &&
                        new Date(prizeBoundStudent.prizeBound.ends_at) > new Date()
          isDisplayableStatus = displayableStatuses.includes(prizeBoundStudent.status)
        }

        return (isNotActivityReward && (hasNoPeriod || isInPeriod || isDisplayableStatus) && isNotQuantity)
      })
    },
    getCombinedPrizeObjects () {
      const prizeBoundStudents = this.participant.participant_info.prize_bound_students
      const prizesBound = this.participant.participant_info.classroom.group.prizes_bound.reduce((prizesBound, prizeBound) => {
        prizesBound[prizeBound.prize_id] = prizeBound
        return prizesBound
      }, {})
      const prizes = this.participant.prizes.reduce((prizes, prize) => {
        prizes[prize.id] = prize
        return prizes
      }, {})
      return prizeBoundStudents.map(prizeBoundStudent => {
        prizeBoundStudent.prizeBound = prizesBound[prizeBoundStudent.prize_id]
        prizeBoundStudent.prize = prizes[prizeBoundStudent.prize_id]
        return prizeBoundStudent
      })
    },
    toggleRewards (id) {
      this.$parent.toggleParticipant(id)
    },
    toggleViewMore () {
      if (this.viewMore) {
        this.viewMore = false
        this.scrollTo()
      } else {
        this.viewMore = true
      }
    },
    scrollTo () {
      setTimeout(() => {
        $('html, body').animate({
          scrollTop: 0,
        }, 300)
      }, 310)
    },
  },
}
</script>
