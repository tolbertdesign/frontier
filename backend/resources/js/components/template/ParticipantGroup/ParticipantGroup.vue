<template>
  <div class="hidden lg:block">
    <!-- Participant Card Groups of 2 -->
    <div :class="['lg:flex flex-wrap', hasOnlyOneParticipant ? 'justify-center' : '' ]">
      <div
        v-for="(participant, index) in participantGroup"
        :key="index"
        :class="[
          selected == null && !hasOnlyOneParticipant ? 'right-border' : '',
          selected == null || selected == participant.id ? '' : 'bg-grey-lighter opacity-75 border-b-2 shadow-inner',
          index == 0 && selected == participant.id && participantGroup.length > 1 ? 'border-r-2' : '',
          index == 1 && selected == participant.id && participantGroup.length > 1 ? 'border-l-2' : ''
        ]"
        class="lg:w-1/2"
      >
        <ParticipantCard
          :participant="participant"
          :program="program"
          :selected="selected"
          @toggle="toggleParticipant($event)"
        />
      </div>
    </div>

    <!-- Participant Rewards Groups of 2 -->
    <div
      v-for="participant in participantGroup"
      :key="participant.id"
      class="px-8"
    >
      <ParticipantRewards
        :program="program"
        :participant="participant"
        :show="selected == participant.id"
      />
    </div>
  </div>
</template>

<script>
import ParticipantCard from '@/components/template/ParticipantCard'
import ParticipantRewards from '@/components/template/ParticipantRewards'

export default {
  name: 'ParticipantGroup',
  components: {
    ParticipantCard,
    ParticipantRewards,
  },
  props: {
    program: {
      type: Object,
      default: undefined,
    },
    participantGroup: {
      type: Array,
      default: () => ([]),
    },
    hasOnlyOneParticipant: {
      type: Boolean,
      default: false,
    },
  },
  data () {
    return {
      selected: null,
    }
  },
  methods: {
    toggleParticipant (id) {
      if (this.selected !== id) {
        this.selected = id
      } else {
        this.selected = null
      }
    },
  },
}
</script>

<style lang="scss">
.right-border:first-of-type {
    position: relative;
    z-index: 1;

    &:before {
        content: "";
        position: absolute;
        right: 0;
        top: 50px;
        height: 8rem;
        width: 1px;
        border-right: 1px solid #DADADA;
    }
}
.top-border {
    position: relative;
    z-index: 1;

    &:before {
        content: "";
        position: absolute;
        margin-left: auto;
        margin-right: auto;
        left: 0;
        right: 0;
        height: 1px;
        width: 80%;
        border-top: 1px solid #DADADA;
    }
}
</style>
