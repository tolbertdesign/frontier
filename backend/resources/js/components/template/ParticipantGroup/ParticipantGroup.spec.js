import { shallowMount, createLocalVue } from '@vue/test-utils'
import ParticipantGroup from './ParticipantGroup.vue'
import ParticipantCard from '@/components/template/ParticipantCard'
import ParticipantRewards from '@/components/template/ParticipantRewards'

const localVue = createLocalVue()

const participant1 = { id: 1 }
const participant2 = { id: 2 }
const program = { id: 9 }

describe('ParticipantGroup', () => {
  it('Snapshop test for ParticipantGroup with one participant', () => {
    const wrapper = shallowMount(ParticipantGroup, {
      localVue,
      components: {
        ParticipantCard,
        ParticipantRewards,
      },
      propsData: {
        program: program,
        participantGroup: [participant1],
        hasOnlyOneParticipant: true,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })
})
describe('ParticipantGroup', () => {
  it('Snapshop test for ParticipantGroup with two participants', () => {
    const wrapper = shallowMount(ParticipantGroup, {
      localVue,
      components: {
        ParticipantCard,
        ParticipantRewards,
      },
      propsData: {
        program: program,
        participantGroup: [participant1, participant2],
        hasOnlyOneParticipant: false,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })
})
