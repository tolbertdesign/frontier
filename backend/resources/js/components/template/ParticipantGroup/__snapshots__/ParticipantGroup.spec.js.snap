// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantGroup Snapshop test for ParticipantGroup with one participant 1`] = `
<div class="hidden lg:block">
  <div class="lg:flex flex-wrap justify-center">
    <div class="lg:w-1/2">
      <participantcard-stub program="[object Object]" participant="[object Object]" group="0"></participantcard-stub>
    </div>
  </div>
  <div class="px-8">
    <participantrewards-stub program="[object Object]" participant="[object Object]"></participantrewards-stub>
  </div>
</div>
`;

exports[`ParticipantGroup Snapshop test for ParticipantGroup with two participants 1`] = `
<div class="hidden lg:block">
  <div class="lg:flex flex-wrap">
    <div class="lg:w-1/2 right-border">
      <participantcard-stub program="[object Object]" participant="[object Object]" group="0"></participantcard-stub>
    </div>
    <div class="lg:w-1/2 right-border">
      <participantcard-stub program="[object Object]" participant="[object Object]" group="0"></participantcard-stub>
    </div>
  </div>
  <div class="px-8">
    <participantrewards-stub program="[object Object]" participant="[object Object]"></participantrewards-stub>
  </div>
  <div class="px-8">
    <participantrewards-stub program="[object Object]" participant="[object Object]"></participantrewards-stub>
  </div>
</div>
`;
